--- a/src/ap/x_snoop.c
+++ b/src/ap/x_snoop.c
@@ -72,7 +72,7 @@ x_snoop_get_l2_packet(struct hostapd_dat
 	struct hostapd_bss_config *conf = hapd->conf;
 	struct l2_packet_data *l2;
 
-	l2 = l2_packet_init(conf->bridge, NULL, ETH_P_ALL, handler, hapd, 1);
+	l2 = l2_packet_init(conf->iface, NULL, ETH_P_ALL, handler, hapd, 1);
 	if (l2 == NULL) {
 		wpa_printf(MSG_DEBUG,
 			   "x_snoop: Failed to initialize L2 packet processing %s",
